### PR TITLE
Fix accidental building of child records when `inverse_of` is specified

### PR DESCRIPTION
**Pull request by clouvet**
### Summary

If an existing record with inverse_of children was duplicated, additional child records would be built. These new records could be persisted upon save or update inadvertently. I added a method to address that case and prevent the accidental creation of additional associations. 

I traced the problem back to [this line in the `replace_on_target` method](https://github.com/clouvet/rails/blob/ecd878bba181e9e89d523e20802f3f46b2d91431/activerecord/lib/active_record/associations/collection_association.rb#L469) in `CollectionAssociation`. Using the steps to reproduce the problem in the original issue as an example, I wrote a test to cover the case of new child records being added to existing owner records when inversing is in play. 

With this change in place, the original steps to reproduce the problem deliver the expected result of no additional records. 

Fixes #44819.

### CODE REVIEW

**Comment by clouvet** [comment]
I force-pushed an update here to resolve an `activerecord/CHANGELOG.md` conflict.

**Comment by clouvet** [comment]
@ghiculescu, since it's been a while I'm giving this a little nudge. I'm happy to revise or close this pull request if this approach isn't preferred.

**Comment by ghiculescu** [comment]
Thanks @clouvet, but this would need review from a member of the Rails core / committers teams. I think the approach is fine. They are pretty busy, but they will eventually get to it :)

It looks like you have a merge conflict btw.

**Comment by clouvet** [comment]
Thanks, @ghiculescu, I appreciate the response. I resolved the conflict with `activerecord/CHANGELOG.md` again.

File: activerecord/CHANGELOG.md
Comment (ghiculescu) [inline activerecord/CHANGELOG.md:12]: I don't think we need this in the changelog, as it's an implementation detail, and not something users will interact with.

File: activerecord/lib/active_record/associations/collection_association.rb
Comment (ghiculescu) [inline activerecord/lib/active_record/associations/collection_association.rb:468]:
```suggestion
            target << record unless new_record_inverse_of_existing_owner?(record, inversing: inversing)
```

**Comment by clouvet** [comment]
@ghiculescu I made changes here in response to your review: I dropped the CHANGELOG addition, accepted your suggestion to use `unless` instead of `if !`, and revised the comment on `new_record_inverse_of_existing_owner?` to explain why that new method exists. 

I'm happy to make additional changes if necessary, and of course I'll squash these multiple commits down to one when the time comes. 

Thanks for the review!

**Comment by clouvet** [comment]
Sorry for the noise here. I squashed the cleanup commits to make this pull request suitable for merging.

File: activerecord/lib/active_record/associations/collection_association.rb
Comment (eileencodes) [inline activerecord/lib/active_record/associations/collection_association.rb:480]: The name is a bit of a mouthful, how about `inverse_of_existing_owner`

```suggestion
        def inverse_of_existing_owner?(record, inversing: false)
```

Also, can you move this under the private namespace, we don't want applications calling this directly.

File: activerecord/test/cases/associations/inverse_associations_test.rb
Comment (eileencodes) [inline activerecord/test/cases/associations/inverse_associations_test.rb:876]: Do you know if there are tests for cases where one of the three `record.new_record? && inversing && !owner.new_record?` isn't true? If not we should add those as well.

**Review by eileencodes — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/associations/collection_association.rb
Comment (clouvet) [reply activerecord/lib/active_record/associations/collection_association.rb]: I renamed the method and moved it to the private namespace.

File: activerecord/test/cases/associations/inverse_associations_test.rb
Comment (clouvet) [reply activerecord/test/cases/associations/inverse_associations_test.rb]: I added more tests that have hopefully more descriptive names to cover cases for when one of `record.new_record? && inversing && !owner.new_record?` is not true. I didn't see existing tests that covered those situations.

The new tests point to a problem though. My proposed solution of checking `inverse_of_existing_owner?` breaks inversing for unsaved objects. This test demonstrates that:
```ruby
  def test_existing_parent_and_new_child_with_inversing_does_not_create_additional_associations
    with_has_many_inversing(Interest) do
      human = Human.create!
      interest = Interest.new(human: human)
      human.update!(name: "name")
      assert_equal human, interest.human
      assert_equal 0, human.interests.size
    end
  end
```

On the current `main` branch, `human.interests.size` should be `1` because of the inversing between the models. So my proposed solution is no good. 

I went around in circles on this point and along the way discovered that `inverse_of` results in autosaving of associated objects. It's unclear to me if it's supposed to, though I looked through the documentation for clarity on that point. 

I pushed up the new commits to show the process I went through in case it's beneficial to solving the original issue, but I imagine this pull request can be closed.

**Pull request closed without merging**
